flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsSubpartitionView.java

```diff
@@ -85,10 +86,12 @@ public BufferAndBacklog getNextBuffer() {
                 updateConsumingStatus(bufferToConsume);
                 return bufferToConsume.map(this::handleBacklog).orElse(null);
             } catch (Throwable cause) {
-                releaseInternal(cause);
-                return null;
+                readError = cause;
             }
         }
+        // release subpartition reader outside of lock to avoid deadlock.
+        releaseInternal(readError);
+        return null;
```

Comment:
We can simply move the `synchronized` block into inside the `try` block.

Reply:
done.